### PR TITLE
Ignore errors on at least one success

### PR DESCRIPTION
**Pull request by storozhukBM**
This change should fix the issue mentioned in https://github.com/cristalhq/hedgedhttp/pull/1
The only thing that is potentially required is randomised exponential back-off between retries.

### CODE REVIEW

File: hedged.go [hunk 1]
```diff
@@ -53,40 +53,49 @@ func (ht *hedgedTransport) RoundTrip(req *http.Request) (*http.Response, error)
 	defer cancel()
 	req = req.WithContext(ctx)
 
-	var res interface{}
-	resultCh := make(chan interface{}, ht.upto)
+	var res *http.Response
+	var err error
+	resultCh := make(chan *http.Response, ht.upto)
```
Comment (cristaloleg) [inline hedged.go:58]: Looks like we're waiting for only 1 results, so the size of then channel can be 1, wdyt?

File: hedged.go [hunk 2]
```diff
@@ -53,40 +53,49 @@ func (ht *hedgedTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
+		select {
+		case res = <-resultCh:
+			break Loop
+		default:
+		}
 
 		select {
 		case res = <-resultCh:
+			break Loop
+		case err = <-errorCh:
```
Comment (cristaloleg) [inline hedged.go:83]: I don't know how bad it's, but we're loosing 1st error, probably we should keep it, wdyt?

(best case is to create a multi-error, but that's too much :)

File: hedged.go [hunk 3]
```diff
@@ -53,40 +53,49 @@ func (ht *hedgedTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
+			continue
 		case <-ctx.Done():
-			res = ctx.Err()
+			err = ctx.Err()
 		case <-time.After(ht.timeout):
```
Comment (cristaloleg) [inline hedged.go:87]: During #1 re-understanding I've realised that we should create `time.Ticker` manually and stop it after the loop, but that's out of scope for this PR, just mentioning for the future.

File: hedged.go [hunk 4]
```diff
@@ -53,40 +53,49 @@ func (ht *hedgedTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
-	case *http.Response:
+	if res != nil {
 		return res, nil
-	default:
-		panic("unreachable")
 	}
+	if err != nil {
+		return nil, err
+	}
+	panic("unreachable")
```
Comment (cristaloleg) [inline hedged.go:98]: This 2 if's and panic can be rewritten as a unconditional switch.

**Review by cristaloleg — APPROVED** [review]

**Comment by cristaloleg** [comment]
Also CC @joe-elliott 'cause this should solve problems with the early errors and cancels.

File: hedged.go (hunk 1 above)
Comment (storozhukBM) [reply hedged.go:58]: Making resultCh with buffer size 1 can lead to leak of goroutines if more than one groutine got successful result.
We can still do it but wait for channel Done() as well.

File: hedged.go (hunk 2 above)
Comment (storozhukBM) [reply hedged.go:83]: Yes, we are going to convey only the last error to the user, not sure if we want to do something more sophisticated than that

Comment (storozhukBM) [reply hedged.go]: I added support for multi-errors

File: hedged.go (hunk 1 above)
Comment (cristaloleg) [reply hedged.go:59]: As I thought, so leaving as it's.

**Review by cristaloleg — APPROVED** [review]

**Pull request merged**